sql/src/main/java/io/crate/planner/consumer/MultiSourceGroupByConsumer.java

```diff
@@ -0,0 +1,231 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.consumer;
+
+import io.crate.analyze.*;
+import io.crate.analyze.relations.AnalyzedRelation;
+import io.crate.analyze.relations.JoinPair;
+import io.crate.analyze.symbol.*;
+import io.crate.collections.Lists2;
+import io.crate.exceptions.UnsupportedFeatureException;
+import io.crate.metadata.Functions;
+import io.crate.metadata.RowGranularity;
+import io.crate.planner.Limits;
+import io.crate.planner.Plan;
+import io.crate.planner.node.ExecutionPhases;
+import io.crate.planner.projection.FilterProjection;
+import io.crate.planner.projection.GroupProjection;
+import io.crate.planner.projection.Projection;
+import io.crate.planner.projection.builder.ProjectionBuilder;
+import io.crate.planner.projection.builder.SplitPoints;
+
+import java.util.*;
+
+public class MultiSourceGroupByConsumer implements Consumer {
+
+    private final Visitor visitor;
+
+    MultiSourceGroupByConsumer(Functions functions) {
+        visitor = new Visitor(functions);
+    }
+
+    @Override
+    public Plan consume(AnalyzedRelation relation, ConsumerContext context) {
+        return visitor.process(relation, context);
+    }
+
+    private static final JoinConditionFieldCollector FIELD_COLLECTOR = new JoinConditionFieldCollector();
```

Comment:
I think we could extract this into a `FieldsVisitor` that has a single static `visitFields(Symbol symbolTree, Consumer<? super Field> consumer)` method. (the same FieldCollector is currently also used in another Consumer)